src/Symfony/Component/JsonCrawler/JsonCrawler.php

```diff
@@ -0,0 +1,401 @@
+                [$left, $right] = array_map('trim', explode($op, $expr, 2));
+                $leftValue = $this->evaluateScalar($left, $context);
+                $rightValue = $this->evaluateScalar($right, $context);
+
+                return $this->compare($leftValue, $rightValue, $op);
+            }
+        }
+
+        if (str_starts_with($expr, '@.')) {
+            $path = substr($expr, 2);
+
+            return \array_key_exists($path, $context);
+        }
+
+        // function calls
+        if (preg_match('/^(\w+)\((.*)\)$/', $expr, $matches)) {
+            $functionName = $matches[1];
+            if (!isset(self::RFC9535_FUNCTIONS[$functionName])) {
+                throw new JsonCrawlerException($expr, \sprintf('invalid function "%s"', $functionName));
+            }
+
+            $functionResult = $this->evaluateFunction($functionName, $matches[2], $context);
+
+            return is_numeric($functionResult) ? $functionResult > 0 : (bool) $functionResult;
+        }
+
+        return false;
+    }
+
+    private function evaluateScalar(string $expr, array $context): mixed
+    {
+        if (is_numeric($expr)) {
+            return str_contains($expr, '.') ? (float) $expr : (int) $expr;
+        }
+
+        if ('true' === $expr) {
+            return true;
+        }
+
+        if ('false' === $expr) {
+            return false;
+        }
+
+        if ('null' === $expr) {
+            return null;
+        }
+
+        // string literals
+        if (preg_match('/^([\'"])(.*)\1$/', $expr, $matches)) {
+            return $matches[2];
+        }
+
+        // current node references
+        if (str_starts_with($expr, '@.')) {
+            $path = substr($expr, 2);
+
+            return $context[$path] ?? null;
+        }
+
+        // function calls
+        if (preg_match('/^(\w+)\((.*)\)$/', $expr, $matches)) {
+            $functionName = $matches[1];
+            if (!isset(self::RFC9535_FUNCTIONS[$functionName])) {
+                throw new JsonCrawlerException($expr, \sprintf('invalid function "%s"', $functionName));
+            }
+
+            return $this->evaluateFunction($functionName, $matches[2], $context);
+        }
+
+        return null;
+    }
+
+    private function evaluateFunction(string $name, string $args, array $context): mixed
+    {
+        $args = array_map(
+            fn ($arg) => $this->evaluateScalar(trim($arg), $context),
+            explode(',', $args)
+        );
+
+        $value = $args[0] ?? null;
+
+        return match ($name) {
+            'length' => match (true) {
+                \is_string($value) => mb_strlen($value),
+                \is_array($value) => \count($value),
+                default => 0,
+            },
+            'count' => \is_array($value) ? \count($value) : 0,
+            'match' => match (true) {
+                \is_string($value) && \is_string($args[1] ?? null) => (bool) @preg_match("/{$args[1]}/", $value),
```

Comment:
should we really silence errors for invalid patterns ?

Reply:
I would say yes, per [this section of the RFC](https://www.rfc-editor.org/rfc/rfc9535.html#section-2.4.6):

> If the first argument is not a string or the second argument is not a string conforming to [[RFC9485](https://www.rfc-editor.org/rfc/rfc9535.html#RFC9485)], the result is LogicalFalse

We could throw something to give a better DX, but it would not follow the specification, if I understood it correctly.